rubix-hadoop2/src/main/java/com/qubole/rubix/hadoop2/Hadoop2ClusterManager.java

```diff
@@ -51,89 +39,41 @@ public void initialize(Configuration conf)
-          {
-            if (!isMaster) {
-              // First time all nodes start assuming themselves as master and down the line figure out their role
-              // Next time onwards, only master will be fetching the list of nodes
-              return ImmutableList.of();
-            }
-            try {
-              List<Hadoop2ClusterManagerUtil.Node> allNodes = Hadoop2ClusterManagerUtil.getAllNodes(yconf);
-              if (allNodes == null) {
-                isMaster = false;
-                return ImmutableList.of();
-              }
-
-              Set<String> hosts = new HashSet<>();
-
-              if (allNodes.isEmpty()) {
-                // Empty result set => server up and only master node running, return localhost has the only node
-                // Do not need to consider failed nodes list as 1node cluster and server is up since it replied to allNodesRequest
-                return ImmutableList.of(InetAddress.getLocalHost().getHostAddress());
-              }
-
-              for (Hadoop2ClusterManagerUtil.Node node : allNodes) {
-                String state = node.getState();
-                log.debug("Hostname: " + node.getNodeHostName() + "State: " + state);
-                //keep only healthy data nodes
-                if (state.equalsIgnoreCase("Running") || state.equalsIgnoreCase("New") || state.equalsIgnoreCase("Rebooted")) {
-                  hosts.add(node.getNodeHostName());
-                }
-              }
-
-              if (hosts.isEmpty()) {
-                throw new Exception("No healthy data nodes found.");
-              }
-
-              List<String> hostList = Lists.newArrayList(hosts.toArray(new String[0]));
-              Collections.sort(hostList);
-              log.debug("Hostlist: " + hostList.toString());
-              return hostList;
-            }
-            catch (Exception e) {
-              throw Throwables.propagate(e);
-            }
-          }
-        }, executor));
-  }
-
-  @Override
-  public boolean isMaster()
-      throws ExecutionException
-  {
-    // issue get on nodesSupplier to ensure that isMaster is set correctly
-    nodesCache.get("nodeList");
-    return isMaster;
   }
 
   @Override
-  public List<String> getNodes()
+  public List<String> getNodesInternal()
   {
     try {
-      return nodesCache.get("nodeList");
+      List<Hadoop2ClusterManagerUtil.Node> allNodes = Hadoop2ClusterManagerUtil.getAllNodes(yconf);
+      if (allNodes == null) {
+        return null;
+      }
+
+      if (allNodes.isEmpty()) {
+        return ImmutableList.of();
+      }
+
+      Set<String> hosts = new HashSet<>();
+      for (Hadoop2ClusterManagerUtil.Node node : allNodes) {
+        String state = node.getState();
+        log.debug("Hostname: " + node.getNodeHostName() + "State: " + state);
+        //keep only healthy data nodes
+        if (state.equalsIgnoreCase("Running") || state.equalsIgnoreCase("New") || state.equalsIgnoreCase("Rebooted")) {
```

Comment:
we may use enums for states.

Reply:
These are not Rubix states but coming from Hadoop. There are enums in Hadoop for NodeState but during serialisation they are converted to string. We can add logic to convert strings back into enums but that should be outside this change (this is existing code).